rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,117 @@
+implementation cost
+* The schema also provides sufficient configurability, e.g
+the generic vendor can be enabled along with nvidia or amd, or it can be
+enabled individually
+
+```bash
+├── cpu/ # CPU code
+├── sycl/ # A common place for basic CPU and GPU SYCL abstractions, e.g. `sycl_cpu_engine_t`, `sycl_gpu_engine_t`, etc.
+└── gpu/ # GPU code. Basic GPU code resides in the GPU directory directly, e.g. `gpu_engine_t`.
+    ├── intel/ # Everything that is related to Intel
+    │   ├── compute/ # Compute layer abstractions
+    │   ├── ocl/ # OpenCL kernels and abstractions (very similar to the current state of `gpu/ocl` directory)
+    │   ├── jit/ # JIT kernels/generators and abstractions (very similar to the current state of `gpu/jit` directory)
+    │   ├── sycl/ # Intel related SYCL functionality/abstractions (e.g. SYCL kernels that use eSIMD, Intel specific extensions, etc)
+    │   └── ...
+    ├── nvidia/ # Everything that is related to NVIDIA
+    │   ├── sycl # NVIDIA related functionality/abstractions (cuDNN/cuBLAS based and CUDA specific SYCL kernels)
+    │   └── ...
+    ├── amd/ # Everything that is related to AMD
+    │   ├── sycl/ # AMD related functionality/abstractions (MIOpen/rocBLAS based and HIP specific SYCL kernels)
+    │   └── ...
+    └── generic/ # Everything that is related to generic kernels
+        └── sycl/ # generic SYCL kernels and related abstractions
+
+```
+
+### Affected Basic Abstractions
+
+The new schema will require moving a lot of parts of the library around
+and while most of the changes are probably just an implementation detail
+there are a few major changes that have to be described in this RFC.
+
+#### Engine
+
+Engine is the most heavily affected abstraction by the changes.
+There is currently no `gpu_engine_t` class as `compute_engine_t` is used as
+an Intel centric alternative of it. In order to have a common basic
+abstraction at the GPU level the `gpu_engine_t` class will be introduced.
+
+The current inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Intel and generic: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_engine_t`
+* SYCL NVIDIA: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_cuda_engine_t`
+* SYCL AMD: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_hip_engine_t`
+* OpenCL (only Intel): `engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
+
+With the new schema they will be changed to:
+* SYCL Intel and generic: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_intel_engine_t`
+* SYCL NVIDIA: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_nvidia_engine_t`
+* SYCL AMD: `engine_t` -> `gpu_engine_t` -> (`gpu_intel_engine_t` -> `compute_engine_t`) -> `sycl_engine_base_t` -> `sycl_gpu_amd_engine_t`
```

Comment:
Does it make sense for AMD and NVidia engines to inherit from engine with Intel in its name - `gpu_intel_engine_t`?

Reply:
I fixed it.